First pass at adding wording for execution::spawn

jesswong opened the pull request:
Loosely based off the wording from P2300? Not totally sure if the exposition-only classes are correct

ispeters committed:
Update [exec.scope.concepts] and tweak the synopsis

This diff is the result of pairing Jessica, Dietmar, and Ruslan; we
updated the formatting of the changes to `[exec.syn]` and tweaked the
wording of `[exec.scope.concepts]`.

jesswong committed:
First pass at execution::spawn wording

dietmarkuehl commented on asyncscope.md:
```diff
@@ -2389,7 +2389,41 @@ creates an asynchronous operation (__[async.ops]__) that, when started:
+[1]{.pnum} `spawn` associates the given input sender with the given token's async scope and, on success, eagerly starts the input sender.
+
+[2]{.pnum} The name `spawn` denotes a customization point object. For subexpressions `sndr`, `token`, and `env`, let `Sndr` be 
+`decltype((sndr))`, let `Token` be `decltype((token)), and let `Env` be `decltype((env))`. If `sender<Sndr>` or `async_scope_token<Snder>` is false, 
```
`Snder` should be `Token`.

ispeters replied:
```suggestion
`decltype((sndr))`, let `Token` be `decltype((token)), and let `Env` be `decltype((env))`. If `sender<Sndr>` or `async_scope_token<Token>` is false, 
```

dietmarkuehl commented on asyncscope.md:
```diff
@@ -2389,7 +2389,41 @@ creates an asynchronous operation (__[async.ops]__) that, when started:
+`decltype((sndr))`, let `Token` be `decltype((token)), and let `Env` be `decltype((env))`. If `sender<Sndr>` or `async_scope_token<Snder>` is false, 
+the expression `spawn(sndr, token, env)` is ill-formed. 
+
+// TODO: when/how do I specify that env is optional?  Or do we get the env from sender-in?
```
[3]{.pnum} The expression `spawn(sndr, token)` is expression-equivalent to `spawn(sndr, token, empty_env{})`.

dietmarkuehl commented on asyncscope.md:
```diff
@@ -2389,7 +2389,41 @@ creates an asynchronous operation (__[async.ops]__) that, when started:
+
+// TODO: when/how do I specify that env is optional?  Or do we get the env from sender-in?
+
+`[3]{.pnum} The exposition-only class template `impls-for` ([exec.snd.general])  is specialized for `spawn_t` as follows:
```
`spawn_t` needs to be defined somewhere.

dietmarkuehl commented:
The current synopsis mentions <code>_spawn-env_</code>. I don't think that is needed: that is covered by <code>_join-env_(_make-env_(env), get_env(rcvr))</code> (I'll remove that from my synopsis PR).

There is also <code>_spawn-receiver_</code> and I think that could be used to describe the behaviour of `spawn(sndr, token, env)` possibly together with a <code>_spawn-state-base_</code>. Maybe something like this although after writing this I noticed that `s` also needs to hold the result of `token.try_associate()` and do something to the sender if that fails:

```
struct spawn-state-base { // exposition-only
    virtual void complete() = 0; // exposition-only
};

struct spawn-receiver { // exposition-only
    spawn-state-base* state; // exposition-only
    void set_value() && noexcept { state->complete(); }
    void set_stopped() && noexcept { state->complete(); }
};
```

[N+1]{.pnum} For the expression `spawn(sndr, token, env)` let <code>_new-sender_</code> be the expression <code>token.wrap(_write-env_(sndr, env))</code> and let `alloc` be the result of `get_allocator(env)` if that expression is well-defined and `std::allocator<void>{}` otherwise. The evaluation of `spawn(sndr, token, env)` creates an operation state `o` using <code>connect(_new-sender_, _spawn-receiver_{&s})</code> whose life-time is managed by an object `s` whose type derives from <code>_spawn-state-base_</code>. Any memory used by `s` is allocated and deallocated using `alloc`. When `s.complete()` is evaluated, all the state is deallocated.

ispeters commented on asyncscope.md:
```diff
@@ -2393,7 +2389,41 @@ creates an asynchronous operation (__[async.ops]__) that, when started:
+::: add
+__`std::execution::spawn` [exec.scope.spawn]__
+
+[1]{.pnum} `spawn` associates the given input sender with the given token's async scope and, on success, eagerly starts the input sender.
```
Since the association could fail, maybe this should say it "attempts" to associate…?
```suggestion
[1]{.pnum} `spawn` attempts to associate the given input sender with the given token's async scope and, on success, eagerly starts the input sender.
```

ispeters commented on asyncscope.md:
```diff
@@ -2393,7 +2389,41 @@ creates an asynchronous operation (__[async.ops]__) that, when started:
+
+// TODO: when/how do I specify that env is optional?  Or do we get the env from sender-in?
+
+`[3]{.pnum} The exposition-only class template `impls-for` ([exec.snd.general])  is specialized for `spawn_t` as follows:
```
I think _`impls-for`_ is a helper for defining a sender and its operation state, so I don't think we need it for `spawn`; instead, we need to define the _`spawn-receiver`_ and _`spawn-env`_ expo-only types and explain what `spawn` does with them.

ispeters committed:
Undo changes to async_scope_token from #37

Revert `async_scope_token_for` to `async_scope_token`.

dietmarkuehl committed:
fix typo

Co-authored-by: Ian Petersen <ispeters@gmail.com>

jesswong commented:
- Fallbacks to allocator in sender-attributes before ultimately falling back to std::allocator
- TODO: Should we need special wording for allocator customization?
- write-env should be outside wrap
- create the receiver before we create the state. how to express?
- algorithm in spawn is always written in terms of allocator but if a fallback allocator is used, don't put it in env

dietmarkuehl commented:
Re point 4 (create receiver before state): I don't think we need to say any of that! The implementer _will_ have to create the receiver before `connect()` is called. In my base class approach that effectively happens when `connect(sndr, spawn-receiver(this))` is called (this is the implementation I'm imagining which would go into the standard):

```
template <typename Sndr, typename Alloc>
struct spawn-state
    : spawn-state-base
    , Alloc
{
    using state_t = decltype(connect(::std::declval<Sndr>(), ::std::declval<spawn-receiver>()));
    state_t state;
    
    spawn-state(Sndr&& sndr, Alloc alloc)
        : spawn-state-base()
        , Alloc(alloc)
        , state(connect(::std::forward<Sndr>(sndr), spawn-receiver(this)))
    {
    }
    void complete() override {
        Alloc alloc(*this);
        ::std::allocator_traits<Alloc>::destroy(alloc, *this);
        ::std::allocator_traits<Alloc>::deallocate(alloc, this);
    }
};
```

Typing it out like this, it seems I could nest the receiver within `spawn-state` and get rid of the `virtual` function. I haven't tried it (nor have I compiled the above) but something like that would work.

dietmarkuehl commented:
re point 2 (special allocator customisation): I think we need to come up with an allocator to use, i.e., that it should be looked for in the `env`, if not there in `get_env(sndr)` (I don't think there is actually a `get_attrs`; there is `get_attrs` in [exec] other than the one in `impls_for` but that _builds_ the environment for the sender rather than getting it), and if not there either then `std::allocator<>()`.

ispeters commented on asyncscope.md:
```diff
@@ -2393,7 +2389,41 @@ creates an asynchronous operation (__[async.ops]__) that, when started:
+[1]{.pnum} `spawn` associates the given input sender with the given token's async scope and, on success, eagerly starts the input sender.
+
+[2]{.pnum} The name `spawn` denotes a customization point object. For subexpressions `sndr`, `token`, and `env`, let `Sndr` be 
+`decltype((sndr))`, let `Token` be `decltype((token)), and let `Env` be `decltype((env))`. If `sender<Sndr>` or `async_scope_token<Snder>` is false, 
```
I think the formatting is a bit wonky because there's a missing backtick here.
```suggestion
`decltype((sndr))`, let `Token` be `decltype((token))`, and let `Env` be `decltype((env))`. If `sender<Sndr>` or `async_scope_token<Snder>` is false, 
```

dietmarkuehl committed:
Update asyncscope.md

Co-authored-by: Ian Petersen <ispeters@gmail.com>

ispeters committed:
Change proposed algorithms to be CPOs

Update `execution.syn` to declare `nest`, `spawn`, and `spawn_future` as
CPOs rather than function templates. I followed the examples in the
`[exec.adapt], sender adaptors` section of the existing
`[execution.syn]` in the working draft.

ispeters committed:
Update the front matter and fix some typos

- Rev the paper number to D3149R7
- Add Dietmar and Ruslan to the contributors list
- Add R7 to the revision history in section 1
- Fix the order of arguments passed to `spawn_future` in section 4.6.2

jesswong committed:
Update wording for spawn with suggestions

jesswong commented on asyncscope.md:
```diff
@@ -2390,40 +2390,126 @@ creates an asynchronous operation (__[async.ops]__) that, when started:
 
-- [3.2]{.pnum} The member `impls-for<spawn_t>::start` is initialized with a callable object equivalent to the following lambda expression:
-// TODO: is this the right exposition only function?
+    @_spawn-state_@(Alloc alloc, Sender sndr, Token token); // see below
```
Section 6 - 9 are inspired by split in P2300: https://eel.is/c++draft/exec#split-5

I'm leaving in section 10 for now in case section 6-9 needs to be redone.

jesswong committed:
Add italics in relevant places and new lines

jesswong committed:
Fix merge conflicts

jesswong committed:
add missing word

rarutyun committed:
Merge branch 'main' into add-spawn-wording
